File: LintCode-Python/mergeKSortedArrays.py
# Given k sorted integer arrays, merge them into one sorted array.

# Have you met this question in a real interview? Yes
# Example
# Given 3 sorted arrays:

# [
#   [1, 3, 5, 7],
#   [2, 4, 6],
#   [0, 8, 9, 10, 11]
# ]
# return [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].

# Challenge
# Do it in O(N log k).

# N is the total number of integers.
# k is the number of arrays.

try:
    import Queue as Q
except ImportError:
    import queue as Q

class Element:
    def __init__(self, row, col, val):
        self.row = row
        self.col = col
        self.val = val
    def __cmp__(self, other):
        return cmp(self.val, other.val) # use the val as metric, the top is min one

def mergeKSortedArray(arrays):
    if not arrays:
        return arrays
    q = Q.PriorityQueue()
    result = []
    # put the first num of each array into q
    for i in range(len(arrays)):
        if len(arrays[i]) > 0:
            elem = Element(i, 0, arrays[i][0])
            q.put(elem)

    # get the min num from q to result
    # each time get the elem from the q, put the next num with same row in q from arrays(if any)
    while not q.empty():
        elem = q.get()
        result.append(elem.val)
        if elem.col < len(arrays[elem.row]) - 1:
            elem = Element(elem.row, elem.col + 1, arrays[elem.row][elem.col + 1])
            q.put(elem)

    return result


print mergeKSortedArray([[1, 3, 5, 7], [2, 4, 6], [0, 8, 9, 10, 11]])

File: LeetCode_Java/src/google_intern/MergeKSortedArrays.java
package google_intern;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * Given k sorted integer arrays, merge them into one sorted array.
 * Example
 * Given 3 sorted arrays:
 * [
 *   [1, 3, 5, 7],
 *   [2, 4, 6],
 *   [0, 8, 9, 10, 11]
 * ]
 * return [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11].
 * Challenge
 * Do it in O(N log k).
 * N is the total number of integers.
 * k is the number of arrays.
 * @author wendi
 *
 */
public class MergeKSortedArrays {
	
	class Node {
		int row, col;
		int val;
		public Node(int row, int col, int val) {
			this.row = row;
			this.col = col;
			this.val = val;
		}
	}
	
	/**
	 * MinHeap
	 * @param int[][] arrays
	 * @return int[]
	 * Time: O(nlog(k))
	 * Space: O(k)
	 */
    public int[] mergeKSortedArrays(int[][] arrays) {
        if (arrays == null || arrays.length == 0) return null;
        int k = arrays.length;
        PriorityQueue<Node> minHeap = new PriorityQueue<>(k, new Comparator<Node>() {
        	@Override
        	public int compare(Node a, Node b) {
        		return a.val - b.val;
        	}
        });
        
        int size = 0;
        for (int i = 0; i < k; i++) {
        	if (arrays[i] == null || arrays[i].length == 0) continue;
        	minHeap.offer(new Node(i, 0, arrays[i][0]));
        	size += arrays[i].length;
        }
        
        int[] res = new int[size];
        for (int i = 0; i < size; i++) {
        	Node curr = minHeap.poll();
        	res[i] = curr.val;
        	if (curr.col != arrays[curr.row].length - 1) {
        		minHeap.offer(new Node(curr.row, curr.col + 1, arrays[curr.row][curr.col + 1]));
        	}
        }
        
        return res;
    }

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		MergeKSortedArrays result = new MergeKSortedArrays();
		System.out.println(Arrays.toString(result.mergeKSortedArrays(new int[][] {{1, 3, 5, 7},{2, 4, 6},{0, 8, 9, 10, 11}})));
	}

}
